src/messaging/ReliableMessageManager.h

```diff
@@ -0,0 +1,113 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the classes corresponding to CHIP reliable message
+ *      protocol.
+ */
+
+#pragma once
+
+#include <stdint.h>
+
+#include <messaging/ReliableMessageProtocolConfig.h>
+
+#include <core/CHIPError.h>
+#include <system/SystemLayer.h>
+#include <system/SystemPacketBuffer.h>
+#include <system/SystemTimer.h>
+
+namespace chip {
+namespace messaging {
+
+class ReliableMessageContext;
+
+class ReliableMessageManager
+{
+private:
+    /**
+     *  @class RetransTableEntry
+     *
+     *  @brief
+     *    This class is part of the CHIP Reliable Messaging Protocol and is used
+     *    to keep track of CHIP messages that have been sent and are expecting an
+     *    acknowledgment back. If the acknowledgment is not received within a
+     *    specific timeout, the message would be retransmitted from this table.
+     *
+     */
+    class RetransTableEntry
+    {
+    public:
+        RetransTableEntry();
+
+        ReliableMessageContext * rc;   /**< The context for the stored CHIP message. */
+        System::PacketBuffer * msgBuf; /**< A pointer to the PacketBuffer object holding the CHIP message. */
+        uint32_t msgId;                /**< The message identifier of the CHIP message awaiting acknowledgment. */
+        uint16_t msgSendFlags;
+        uint16_t nextRetransTimeTick; /**< A counter representing the next retransmission time for the message. */
+        uint8_t sendCount;            /**< A counter representing the number of times the message has been sent. */
+    };
+
+public:
+    ReliableMessageManager();
+    ~ReliableMessageManager();
+
+    void FreeContext(ReliableMessageContext * rc);
+
+    uint64_t GetTickCounterFromTimePeriod(uint64_t period);
+    uint64_t GetTickCounterFromTimeDelta(uint64_t newTime);
+
+    void ExecuteActions();
+    void ProcessDDMessage(uint32_t PauseTimeMillis, uint64_t DelayedNodeId);
+    static void Timeout(System::Layer * aSystemLayer, void * aAppState, System::Error aError);
+
+    CHIP_ERROR AddToRetransTable(ReliableMessageContext * rc, System::PacketBuffer * msgBuf, uint32_t messageId,
+                                 uint16_t msgSendFlags, RetransTableEntry ** rEntry);
+    void PauseRetransTable(ReliableMessageContext * rc, uint32_t PauseTimeMillis);
+    void ResumeRetransTable(ReliableMessageContext * rc);
+    bool CheckAndRemRetransTable(ReliableMessageContext * rc, uint32_t msgId);
+    CHIP_ERROR SendFromRetransTable(RetransTableEntry * entry);
+    void ClearRetransmitTable(ReliableMessageContext * rc);
+    void ClearRetransmitTable(RetransTableEntry & rEntry);
+    void FailRetransmitTableEntries(ReliableMessageContext * rc, CHIP_ERROR err);
+
+    void StartTimer();
+    void StopTimer();
+    void ExpireTicks();
+
+    ReliableMessageProtocolConfig mConfig; /**< ReliableMessageProtocol configuration. */
```

Comment:
data members should be private.

Unsure if doccomment is required ('ReliableMessageProtocolConfig' is obviously a 'ReliableMessageProtocol configuration').